File: src/rust/ir.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum IRType {
    LOAD_CONST,
    LOAD_VAR,
    STORE_VAR,
    CALL,
    FFI_GET_ATTR,
    FFI_CALL,
    FFI_IMPORT,
    RETURN,
    JUMP,
    JUMP_IF_FALSE,
    JUMP_IF_TRUE,
    LOOP,
    LOOP_BEGIN,
    LOOP_END,
    ITER,
    ITER_INIT,
    ALLOC,
    FREE,
    INCREMENT_RC,
    DECREMENT_RC,
    SPAWN_THREAD,
    WAIT,
    YIELD,
    AWAIT,
    BINARY_OP,
    UNARY_OP,
    NOP,
    LABEL,
    DEFER,
    MATCH,
    TRY,
}

impl Default for IRType {
    fn default() -> Self {
        IRType::NOP
    }
}

#[derive(Debug, Clone, Default)]
pub struct IRInstruction {
    pub ty: IRType,
    pub operand: Option<String>,
    pub arg1: Option<String>,
    pub arg2: Option<String>,
    pub arg3: Option<String>,
}

impl IRInstruction {
    pub fn new(ty: IRType) -> Self {
        IRInstruction { ty, operand: None, arg1: None, arg2: None, arg3: None }
    }

    pub fn with_operand(ty: IRType, operand: impl Into<String>) -> Self {
        IRInstruction { ty, operand: Some(operand.into()), arg1: None, arg2: None, arg3: None }
    }
}

#[derive(Debug, Clone)]
pub struct IRModule {
    pub name: String,
    pub functions: HashMap<String, Vec<IRInstruction>>,
    pub variables: HashMap<String, Option<String>>,
    pub imports: HashMap<String, (String, String)>,
    pub annotation_registry: HashMap<String, String>,
    pub func_params: HashMap<String, Vec<String>>,
}

impl Default for IRModule {
    fn default() -> Self {
        IRModule {
            name: "main".to_string(),
            functions: HashMap::new(),
            variables: HashMap::new(),
            imports: HashMap::new(),
            annotation_registry: HashMap::new(),
            func_params: HashMap::new(),
        }
    }
}

impl IRModule {
    pub fn add_function(&mut self, name: &str, instructions: Vec<IRInstruction>) {
        self.functions.insert(name.to_string(), instructions);
    }

    pub fn add_variable(&mut self, name: &str, type_str: Option<&str>) {
        self.variables.insert(name.to_string(), type_str.map(|s| s.to_string()));
    }
}

pub fn load_const(value: &str) -> IRInstruction { IRInstruction::with_operand(IRType::LOAD_CONST, value) }
pub fn load_var(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::LOAD_VAR, name) }
pub fn store_var(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::STORE_VAR, name) }
pub fn call(func: &str, _args: Vec<String>) -> IRInstruction { IRInstruction::with_operand(IRType::CALL, func) }
pub fn ret() -> IRInstruction { IRInstruction::new(IRType::RETURN) }
pub fn jump(label: &str) -> IRInstruction { IRInstruction::with_operand(IRType::JUMP, label) }
pub fn jump_if_false(label: &str) -> IRInstruction { IRInstruction::with_operand(IRType::JUMP_IF_FALSE, label) }
pub fn jump_if_true(label: &str) -> IRInstruction { IRInstruction::with_operand(IRType::JUMP_IF_TRUE, label) }
pub fn loop_begin(label: &str) -> IRInstruction { IRInstruction::with_operand(IRType::LOOP_BEGIN, label) }
pub fn loop_end(label: &str) -> IRInstruction { IRInstruction::with_operand(IRType::LOOP_END, label) }
pub fn alloc(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::ALLOC, name) }
pub fn free(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::FREE, name) }
pub fn increment_rc(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::INCREMENT_RC, name) }
pub fn decrement_rc(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::DECREMENT_RC, name) }
pub fn spawn_thread(func: &str, _args: Vec<String>) -> IRInstruction { IRInstruction::with_operand(IRType::SPAWN_THREAD, func) }
pub fn wait(task_id: &str) -> IRInstruction { IRInstruction::with_operand(IRType::WAIT, task_id) }
pub fn yield_op() -> IRInstruction { IRInstruction::new(IRType::YIELD) }
pub fn await_op(func: &str) -> IRInstruction { IRInstruction::with_operand(IRType::AWAIT, func) }
pub fn binary_op(op: &str) -> IRInstruction { IRInstruction::with_operand(IRType::BINARY_OP, op) }
pub fn unary_op(op: &str) -> IRInstruction { IRInstruction::with_operand(IRType::UNARY_OP, op) }
pub fn iter_init(var_name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::ITER_INIT, var_name) }
pub fn ffi_import(language: &str, module: &str, as_name: &str) -> IRInstruction {
    IRInstruction { ty: IRType::FFI_IMPORT, operand: Some(language.to_string()), arg1: Some(module.to_string()), arg2: Some(as_name.to_string()), arg3: None }
}
pub fn ffi_get_attr(module: &str, attr: &str) -> IRInstruction {
    IRInstruction { ty: IRType::FFI_GET_ATTR, operand: Some("python".to_string()), arg1: Some(module.to_string()), arg2: Some(attr.to_string()), arg3: None }
}
pub fn ffi_call(language: &str, module: &str, func: &str, _arg_count: i64) -> IRInstruction {
    IRInstruction { ty: IRType::FFI_CALL, operand: Some(language.to_string()), arg1: Some(module.to_string()), arg2: Some(func.to_string()), arg3: None }
}
pub fn label(name: &str) -> IRInstruction { IRInstruction::with_operand(IRType::LABEL, name) }
pub fn defer_op(expr: &str) -> IRInstruction { IRInstruction::with_operand(IRType::DEFER, expr) }
pub fn match_op() -> IRInstruction { IRInstruction::new(IRType::MATCH) }
pub fn try_op() -> IRInstruction { IRInstruction::new(IRType::TRY) }

File: src/synthlang/ir.py
"""SynthLang Intermediate Representation - Linear IR for the VM."""
from enum import Enum, auto
from dataclasses import dataclass
from typing import Any, Optional, Dict, List


class IRType(Enum):
    LOAD_CONST = auto()
    LOAD_VAR = auto()
    STORE_VAR = auto()
    CALL = auto()
    FFI_GET_ATTR = auto()
    FFI_CALL = auto()
    FFI_IMPORT = auto()
    RETURN = auto()
    JUMP = auto()
    JUMP_IF_FALSE = auto()
    JUMP_IF_TRUE = auto()
    LOOP = auto()
    LOOP_BEGIN = auto()
    LOOP_END = auto()
    ITER = auto()
    ITER_INIT = auto()
    ALLOC = auto()
    FREE = auto()
    INCREMENT_RC = auto()
    DECREMENT_RC = auto()
    SPAWN_THREAD = auto()
    WAIT = auto()
    YIELD = auto()
    AWAIT = auto()
    BINARY_OP = auto()
    UNARY_OP = auto()
    NOP = auto()
    LABEL = auto()
    DEFER = auto()
    MATCH = auto()
    TRY = auto()
    EXEC_CODE_BLOCK = auto()
    MAKE_CHANNEL = auto()
    CHAN_SEND = auto()
    CHAN_RECV = auto()
    BUILD_KWARGS = auto()


@dataclass
class IRInstruction:
    type: IRType
    operand: Any = None
    arg1: Any = None
    arg2: Any = None
    arg3: Any = None
    result: Any = None

    def __repr__(self):
        parts = [f"{self.type.name}"]
        if self.operand is not None:
            parts.append(f"operand={self.operand}")
        if self.arg1 is not None:
            parts.append(f"arg1={self.arg1}")
        if self.arg2 is not None:
            parts.append(f"arg2={self.arg2}")
        if self.arg3 is not None:
            parts.append(f"arg3={self.arg3}")
        if self.result is not None:
            parts.append(f"result={self.result}")
        return f"IR({', '.join(parts)})"


class IRModule:
    def __init__(self, name: str = "main"):
        self.name = name
        self.functions: dict = {}
        self.variables: dict = {}
        self.imports: Dict[str, tuple] = {}
        self.annotations: list = []
        self.func_params: Dict[str, List[str]] = {}
        self.annotation_registry: Dict[str, Any] = {}

    def add_function(self, name: str, instructions: list):
        self.functions[name] = instructions

    def add_variable(self, name: str, type_str: str = None):
        self.variables[name] = type_str

    def register_annotation(self, annotation_name: str, target: str):
        self.annotation_registry[annotation_name] = target

    def __repr__(self):
        return f"IRModule({self.name}, {len(self.functions)} functions, {len(self.variables)} variables)"


def load_const(value: Any) -> IRInstruction:
    return IRInstruction(IRType.LOAD_CONST, operand=value)


def load_var(name: str) -> IRInstruction:
    return IRInstruction(IRType.LOAD_VAR, operand=name)


def store_var(name: str) -> IRInstruction:
    return IRInstruction(IRType.STORE_VAR, operand=name)


def call(func: str, args: list) -> IRInstruction:
    return IRInstruction(IRType.CALL, operand=func, arg1=args)


def ret() -> IRInstruction:
    return IRInstruction(IRType.RETURN)


def jump(label: str, result: str = None) -> IRInstruction:
    return IRInstruction(IRType.JUMP, operand=label, result=result)


def jump_if_false(label: str) -> IRInstruction:
    return IRInstruction(IRType.JUMP_IF_FALSE, operand=label)


def jump_if_true(label: str) -> IRInstruction:
    return IRInstruction(IRType.JUMP_IF_TRUE, operand=label)


def loop_begin(label: str) -> IRInstruction:
    return IRInstruction(IRType.LOOP_BEGIN, operand=label)


def loop_end(label: str) -> IRInstruction:
    return IRInstruction(IRType.LOOP_END, operand=label)


def alloc(name: str) -> IRInstruction:
    return IRInstruction(IRType.ALLOC, operand=name)


def free(name: str) -> IRInstruction:
    return IRInstruction(IRType.FREE, operand=name)


def increment_rc(name: str) -> IRInstruction:
    return IRInstruction(IRType.INCREMENT_RC, operand=name)


def decrement_rc(name: str) -> IRInstruction:
    return IRInstruction(IRType.DECREMENT_RC, operand=name)


def spawn_thread(func: str, args: list) -> IRInstruction:
    return IRInstruction(IRType.SPAWN_THREAD, operand=func, arg1=args)


def wait(task_id: int) -> IRInstruction:
    return IRInstruction(IRType.WAIT, operand=task_id)


def yield_op() -> IRInstruction:
    return IRInstruction(IRType.YIELD)


def await_op(func: str) -> IRInstruction:
    return IRInstruction(IRType.AWAIT, operand=func)


def binary_op(op: str) -> IRInstruction:
    return IRInstruction(IRType.BINARY_OP, operand=op)


def unary_op(op: str) -> IRInstruction:
    return IRInstruction(IRType.UNARY_OP, operand=op)


def iter_init(var_name: str) -> IRInstruction:
    return IRInstruction(IRType.ITER_INIT, operand=var_name)


def ffi_import(language: str, module: str, as_name: str = None) -> IRInstruction:
    return IRInstruction(IRType.FFI_IMPORT, operand=language, arg1=module, arg2=as_name)


def ffi_get_attr(module: str, attr: str) -> IRInstruction:
    return IRInstruction(IRType.FFI_GET_ATTR, operand='python', arg1=module, arg2=attr)


def ffi_call(language: str, module: str, func: str, args: list) -> IRInstruction:
    return IRInstruction(IRType.FFI_CALL, operand=language, arg1=module, arg2=func, arg3=args)


def label(name: str) -> IRInstruction:
    return IRInstruction(IRType.LABEL, operand=name)

def defer_op(expr: str) -> IRInstruction:
    return IRInstruction(IRType.DEFER, operand=expr)

def match_op(cases: list) -> IRInstruction:
    return IRInstruction(IRType.MATCH, operand=str(cases))

def try_op() -> IRInstruction:
    return IRInstruction(IRType.TRY)

def exec_code_block(lang: str, code: str) -> IRInstruction:
    return IRInstruction(IRType.EXEC_CODE_BLOCK, operand=lang, arg1=code)

def make_channel(size: int = 0) -> IRInstruction:
    return IRInstruction(IRType.MAKE_CHANNEL, operand=size)

def chan_send(chan_name: str) -> IRInstruction:
    return IRInstruction(IRType.CHAN_SEND, operand=chan_name)

def chan_recv(chan_name: str) -> IRInstruction:
    return IRInstruction(IRType.CHAN_RECV, operand=chan_name)

def build_kwargs(keys: list) -> IRInstruction:
    return IRInstruction(IRType.BUILD_KWARGS, operand=keys)